File: frontend/src/mocks/handlers/schemas.ts
/**
 * Schema API handlers
 */

import { http, HttpResponse, delay } from 'msw'
import { getStore, getById, getSchemaBySourceId } from '../data/store'
import { createSchema, createId } from '../factories'

const API_BASE = '/api/v1'

export const schemasHandlers = [
  // Get schema for a source
  http.get(`${API_BASE}/sources/:sourceId/schema`, async ({ params }) => {
    await delay(150)

    const sourceId = params.sourceId as string
    const source = getById(getStore().sources, sourceId)

    if (!source) {
      return HttpResponse.json(
        { detail: 'Source not found' },
        { status: 404 }
      )
    }

    const schema = getSchemaBySourceId(sourceId)

    if (!schema) {
      return HttpResponse.json(null)
    }

    return HttpResponse.json(schema)
  }),

  // Learn schema for a source
  http.post(`${API_BASE}/sources/:sourceId/learn`, async ({ params }) => {
    await delay(1000) // Simulate learning time

    const sourceId = params.sourceId as string
    const source = getById(getStore().sources, sourceId)

    if (!source) {
      return HttpResponse.json(
        { detail: 'Source not found' },
        { status: 404 }
      )
    }

    // Create or update schema
    const schema = createSchema({
      id: createId(),
      sourceId,
    })

    getStore().schemas.set(sourceId, schema)

    // Update source's has_schema flag
    const sources = getStore().sources
    sources.set(sourceId, { ...source, has_schema: true })

    return HttpResponse.json(schema)
  }),

  // Update schema
  http.put(`${API_BASE}/sources/:sourceId/schema`, async ({ params, request }) => {
    await delay(300)

    const sourceId = params.sourceId as string
    const source = getById(getStore().sources, sourceId)

    if (!source) {
      return HttpResponse.json(
        { detail: 'Source not found' },
        { status: 404 }
      )
    }

    const body = (await request.json()) as { schema_yaml: string }

    const existingSchema = getSchemaBySourceId(sourceId)

    const updatedSchema = {
      ...(existingSchema ?? createSchema({ sourceId })),
      schema_yaml: body.schema_yaml,
      updated_at: new Date().toISOString(),
    }

    getStore().schemas.set(sourceId, updatedSchema)

    return HttpResponse.json(updatedSchema)
  }),
]

File: src/truthound_dashboard/api/schemas.py
"""Schemas API endpoints.

This module provides endpoints for schema learning and management.
"""

from __future__ import annotations

from typing import Annotated

from fastapi import APIRouter, HTTPException, Path

from truthound_dashboard.schemas import (
    SchemaLearnRequest,
    SchemaResponse,
    SchemaUpdateRequest,
)

from .deps import SchemaServiceDep, SourceServiceDep

router = APIRouter()


@router.get(
    "/sources/{source_id}/schema",
    response_model=SchemaResponse | None,
    summary="Get source schema",
    description="Get the active schema for a data source",
)
async def get_schema(
    service: SchemaServiceDep,
    source_service: SourceServiceDep,
    source_id: Annotated[str, Path(description="Source ID")],
) -> SchemaResponse | None:
    """Get the active schema for a source.

    Args:
        service: Injected schema service.
        source_service: Injected source service.
        source_id: Source to get schema for.

    Returns:
        Schema if exists, None otherwise.

    Raises:
        HTTPException: 404 if source not found.
    """
    # Verify source exists
    source = await source_service.get_by_id(source_id)
    if source is None:
        raise HTTPException(status_code=404, detail="Source not found")

    schema = await service.get_schema(source_id)
    if schema is None:
        return None
    return SchemaResponse.from_model(schema)


@router.post(
    "/sources/{source_id}/learn",
    response_model=SchemaResponse,
    summary="Learn schema",
    description="Auto-learn schema from data source",
)
async def learn_schema(
    service: SchemaServiceDep,
    source_service: SourceServiceDep,
    source_id: Annotated[str, Path(description="Source ID")],
    request: SchemaLearnRequest,
) -> SchemaResponse:
    """Learn schema from a data source.

    Args:
        service: Injected schema service.
        source_service: Injected source service.
        source_id: Source to learn schema from.
        request: Learning options.

    Returns:
        Learned schema.

    Raises:
        HTTPException: 404 if source not found.
    """
    # Verify source exists
    source = await source_service.get_by_id(source_id)
    if source is None:
        raise HTTPException(status_code=404, detail="Source not found")

    try:
        schema = await service.learn_schema(
            source_id,
            infer_constraints=request.infer_constraints,
        )
        return SchemaResponse.from_model(schema)
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))


@router.put(
    "/sources/{source_id}/schema",
    response_model=SchemaResponse,
    summary="Update schema",
    description="Update the schema YAML for a source",
)
async def update_schema(
    service: SchemaServiceDep,
    source_service: SourceServiceDep,
    source_id: Annotated[str, Path(description="Source ID")],
    request: SchemaUpdateRequest,
) -> SchemaResponse:
    """Update schema YAML for a source.

    Args:
        service: Injected schema service.
        source_service: Injected source service.
        source_id: Source to update schema for.
        request: New schema YAML.

    Returns:
        Updated schema.

    Raises:
        HTTPException: 404 if source or schema not found.
    """
    # Verify source exists
    source = await source_service.get_by_id(source_id)
    if source is None:
        raise HTTPException(status_code=404, detail="Source not found")

    schema = await service.update_schema(source_id, request.schema_yaml)
    if schema is None:
        raise HTTPException(
            status_code=404,
            detail="No active schema found. Use /learn to create one first.",
        )
    return SchemaResponse.from_model(schema)
